Add TSFE documentation (#2736)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/TSFE/Index.rst

is page rendering. It also handles reading from and writing to the page cache.
For more details it is best to look into the source code.

There are several contexts in which the term TSFE is used:

*   PHP: It is passed as request attribute
    :ref:`frontend.controller <typo3-request-attribute-frontend-controller>`
*   PHP: It was and is available as global array :php:`$GLOBALS['TSFE']` in PHP.
*   TypoScript: TypoScript function :ref:`TSFE <t3tsref:data-type-gettext-tsfe>`
    which can be used to access public properties in TSFE.
*   (deprecated since v9.5 and :doc:`removed in 10.0
    <ext_core:Changelog/10.0/Breaking-88564-PageTSconfigSettingTSFEconstantsRemoved>`)
    Page TSconfig: :typoscript:`TSFE.constants`.

The TypoScript part is covered in the
:ref:`TypoScript Reference: TSFE <t3tsref:data-type-gettext-tsfe>`.
In this section we focus on the PHP part and give an overview, in which way the
TSFE class can be used.

Accessing TSFE
===============

.. attention::

    Some of the former public properties and methods have been changed to
    protected or marked as internal. Often, accessing TSFE is no longer
    necessary, and there are better alternatives.

>     Access :php:`$GLOBALS['TSFE']` directly only as a last resort.

Reviewer: ```suggestion
    Access :php:`$GLOBALS['TSFE']` directly only as a last resort,
```

>     strongly discouraged if not absolutely necessary.

Reviewer: ```suggestion
    usage is strongly discouraged, if not absolutely necessary.
```

.. code-block:: php

    // TYPO3\CMS\Core\Site\Entity\SiteLanguage object.
    $language = $request->getAttribute('language');
    $languageId = $language->getLanguageId();

If the request is not available, accessing language settings
can be done using the :ref:`language aspect <context_api_aspects_language>`.

Get the language of the current page as integer:

.. code-block:: php

    $languageId = (int) $context->getPropertyFromAspect('language', 'id');

.. _tsfe_frontendUser:

Access frontend user information
--------------------------------

.. code-block:: php

    // !!! discouraged
    $feUser = $GLOBALS['TSFE']->fe_user;

Use the :ref:`frontend.user <typo3-request-attribute-frontend-user>`:

.. code-block:: php

    // TYPO3\CMS\Frontend\Authentication\FrontendUserAuthentication
    $frontendUser = $request->getAttribute('frontend.user');

Some information via frontend and backend users con be obtained via the
:ref:`user aspect <context_api_aspects_user>`. For example:

.. code-block:: php

    // return whether a frontend user is logged in
    $context->getPropertyFromAspect('frontend.user', 'isLoggedIn');

.. _tsfe_baseURL:

Get current base URL
--------------------

> It used to be possible to get the baseUrl configuration (from TypoScript

Reviewer: ```suggestion
It used to be possible to get the base URL configuration (from TypoScript
```

> config.baseURL) with the :php:`TSFE` :php:`baseURL` property. The

Reviewer: ```suggestion
:typoscript:`config.baseURL`) with the :php:`TSFE` :php:`baseURL` property. The
```

.. include:: /Includes.rst.txt
.. index:: TSFE; TypoScriptFrontendController
.. _tsfe:

====

> TSFE

Author: I think "TypoScriptFrontendController (TSFE)" is very long. I am thinking about the menu ... Otherwise it would be better to use the long form, I think.

> TSFE

Reviewer 2: In our new Rendering Container we will be able to use different links for the menu and the main title. This is not possible yet

> TSFE

Reviewer 2: For now I would suggest to go with the long title. After switching the renderings we should generally overhaul the menu tree

====

> 

Reviewer: ```suggestion

..  contents::
    :local:

```

Adding a TOC makes it easier to grasp what this page provides.

What is TSFE?
=============

TSFE is short for :php:`\TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController`,
a class which exists in the system extension EXT:frontend.

As the name implies: A responsibility of TSFE
is page rendering. It also handles reading from and writing to the page cache.

> For more details it is best to look in the source code.

Reviewer: ```suggestion
For more details it is best to look into the source code.
```

There are several contexts in which the term TSFE is used:

> 1.  PHP: It is passed as request attribute

Reviewer: I suggest an unordered list, as the points have no order.

Author: done

    :ref:`frontend.controller <typo3-request-attribute-frontend-controller>`
2.  PHP: It was and is available as global array :php:`$GLOBALS['TSFE']` in PHP.
3.  TypoScript: TypoScript function :ref:`TSFE <t3tsref:data-type-gettext-tsfe>`
    which can be used to access public properties in TSFE.
4.  (deprecated since v9.5 and :doc:`removed in 10.0
    <ext_core:Changelog/10.0/Breaking-88564-PageTSconfigSettingTSFEconstantsRemoved>`)
    Page TSconfig: :typoscript:`TSFE.constants`.

> Focusing on the PHP part (as the TypoScript part is covered in the

Reviewer: To break up this long sentence, I suggest to move the part in brackets into a separate sentence at the end of this one.

Author: done

:ref:`TypoScript Reference: TSFE <t3tsref:data-type-gettext-tsfe>` page),

> this page gives an overview, what can still be used, what is deprecated,

Reviewer: ```suggestion
this page gives an overview, what can be used, what is deprecated,
```
The "still" implies (at least for me), that TSFE is phased-out (which is not true AFAIK).

removed or discouraged and in which way the TSFE class itself may be
interesting to developers.

Accessing TSFE
===============

.. attention::

    Some of the former public properties and methods have been changed to
    protected or marked as internal. Often, accessing TSFE is no longer
    necessary, and there are better alternatives.

>     Directly access :php:`$GLOBALS['TSFE']` only as a last resort. It is

Reviewer: ```suggestion
    Access :php:`$GLOBALS['TSFE']` directly only as a last resort.
```

    strongly discouraged if not absolutely necessary.

From the source:

    When calling a frontend page, an instance of this object is available
    as :php:`$GLOBALS['TSFE']`, even though the Core development strives to get
    rid of this in the future.

If access to the
:php:`\TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController` instance is
necessary, use the request attribute
:ref:`frontend.controller <typo3-request-attribute-frontend-controller>`:

..  code-block:: php

    $frontendController = $request->getAttribute('frontend.controller');

> 

Reviewer: ```suggestion

..  seealso::
    :ref:`getting-typo3-request-object`

```

This links to the page which describes how to get the request object in different contexts.

TSFE is not available in all contexts. In particular, it is
only available in frontend contexts, not in the backend or CLI.

Initializing :php:`$GLOBALS['TSFE']` in the backend is sometimes done in code
examples found online. This is not recommended. TSFE is not initialized in the
backend context by the Core (and there is usually no need to do this).

From the PHP documentation:

    As of PHP 8.1.0, $GLOBALS is now a read-only copy of the global symbol table.
    That is, global variables cannot be modified via its copy.

https://www.php.net/manual/en/reserved.variables.globals.php

Howtos
======

.. add howtos for common use cases, remove the not recommended, deprecated
.. way if there is a better way

Following are some examples which use TSFE and alternatives to using TSFE,
where available:

.. _tsfe_ContentObjectRenderer:

Access ContentObjectRenderer
----------------------------

> Access the :php:`ContentObjectRenderer` (often referred to as "cObj"):

Reviewer: ```suggestion
Access the :php:`\TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer` (often referred to as "cObj"):
```

.. code-block:: php

    // this is discouraged, obtain TSFE from request attribute 'frontend.controller'
    // see next example
    $cObj = $GLOBALS['TSFE']->cObj;

.. code-block:: php

    $frontendController = $request->getAttribute('frontend.controller');
    $cObj = $frontendController->cObj;

> In the case of a non Extbase plugin via setter injection:

Reviewer: ```suggestion
In the case of :ref:`user function <tsref:cobj-user-int>` (for example, a non-Extbase plugin) via setter injection:
```

.. code-block:: php

    public function setContentObjectRenderer(ContentObjectRenderer $cObj): void
    {
        $this->cObj = $cObj;
    }

.. _tsfe_pageId:

> Access current page id

Reviewer: ```suggestion
Access current page ID
```

----------------------

> Access the current page id:

Reviewer: ```suggestion
Access the current page ID:
```

.. code-block:: php

    // this is discouraged

>     $GLOBALS['TSFE']->id

Reviewer: ```suggestion
    $pageId = $GLOBALS['TSFE']->id;
```

Can be done using the :ref:`'routing' <typo3-request-attribute-routing>`
request attribute:

.. code-block:: php

    $pageArguments = $request->getAttribute('routing');
    $pageId = $pageArguments->getPageId();

.. _tsfe_language:

> Access language settings

Reviewer: I am missing here the discouraged usage of TSFE to obtain the language settings. This may help users migrating from older versions.

------------------------

> In order to get current language settings, such as the current language id,

Reviewer: ```suggestion
In order to get current language settings, such as the current language ID,
```

obtain :php:`\TYPO3\CMS\Core\Site\Entity\SiteLanguage` object from the
:ref:`request attribute <request-attributes>` 'language':

.. code-block:: php

    // TYPO3\CMS\Core\Site\Entity\SiteLanguage object.
    $language = $request->getAttribute('language');
    $languageId = $language->getLanguageId();

If the request is not available, accessing language settings
can be done using the :ref:`language aspect <context_api_aspects_language>`.

Get the language of the current page as integer:

.. code-block:: php

    $languageId = (int) $context->getPropertyFromAspect('language', 'id');

.. _tsfe_frontendUser

Access frontend user information
--------------------------------

> Accessing information about Frontend users. Accessing

Reviewer: ```suggestion
Accessing information about frontend users. Accessing
```

:php:`$GLOBALS['TSFE']->fe_user` directly is discouraged.

.. code-block:: php

>     // not recommended, use alternatives if possible

Reviewer: ```suggestion
    // discouraged, use alternatives if possible
```

We should use the same wording on this page.

    // TYPO3\CMS\Frontend\Authentication\FrontendUserAuthentication
    $feUser = $GLOBALS['TSFE']->fe_user;

> Using :ref:`request-attributes`:

Reviewer: ```suggestion
Using the request attribute :ref:`frontend.user <typo3-request-attribute-frontend-user>`:
```

.. code-block:: php

    // TYPO3\CMS\Frontend\Authentication\FrontendUserAuthentication
    $frontendUser = $request->getAttribute('frontend.user');

Some information via frontend and backend users con be obtained via the
:ref:`user aspect <context_api_aspects_user>`. For example:

.. code-block:: php

    // return whether a frontend user is logged in
    $context->getPropertyFromAspect('frontend.user', 'isLoggedIn');

.. _tsfe_baseURL:

Get current base URL
--------------------

> To get the base URL of the current site (or other site configuration), use site

Reviewer: I am missing the discouraged usage of TSFE here also.

Author: done

> To get the base URL of the current site (or other site configuration), use site

Reviewer: ```suggestion
To get the base URL of the current site (or other :ref:`site configuration <sitehandling>`), use site
```

configuration:

.. code-block:: php

    // TYPO3\CMS\Core\Site\Entity\Site
    $site = $request->getAttribute('site');
    // array
    $siteConfiguration = $site->getConfiguration();
    $baseUrl = $siteConfiguration['base'];

.. _tsfe_siteByPageId:

> Get site by page id

Reviewer: ```suggestion
Get site by page ID
```

I am also missing the discouraged usage of TSFE.

Author: Not sure why you capitalize id here.  I think both can we used or is this a convention?

Using the request attribute :ref:`frontend.user <typo3-request-attribute-frontend-user>`:

.. code-block:: php

    // TYPO3\CMS\Frontend\Authentication\FrontendUserAuthentication
    $frontendUser = $request->getAttribute('frontend.user');

Some information via frontend and backend users con be obtained via the
:ref:`user aspect <context_api_aspects_user>`. For example:

.. code-block:: php

    // return whether a frontend user is logged in
    $context->getPropertyFromAspect('frontend.user', 'isLoggedIn');

.. _tsfe_baseURL:

Get current base URL
--------------------

To get the base URL of the current site (or other :ref:`site configuration <sitehandling>`), use site
configuration:

.. code-block:: php

    // It used to be possible to get the baseUrl configuration (from TypoScript
    // config.baseURL) this way.
    // The property is now protected and deprecated since TYPO3 v12,
    // use site configuration
    $GLOBALS['TSFE']->baseURL

.. code-block:: php

    // TYPO3\CMS\Core\Site\Entity\Site
    $site = $request->getAttribute('site');
    // array
    $siteConfiguration = $site->getConfiguration();
    $baseUrl = $siteConfiguration['base'];

.. _tsfe_siteByPageId:

> Get site by page ID

Reviewer: I am missing the relation of site to TSFE. This is an example how to get the site. Is/was it possible with TSFE to get the site?

Reviewer 3: yes `TSFE->id` or `$GLOBALS['TSFE']->id`

Reviewer: Hmm, this is the page ID, not getting the site through the page ID?

The section should point to it: TSFE (discouraged), SiteFinder (the way to go).

Reviewer 3: oh sorry.
I don't know how to get an arbitrary site by page ID.  
Just the current page was something like  `$GLOBALS['TSFE']->page`, don't remember exactly

Reviewer: Ah, there is a public TSFE->getSite() method. This should be added as discouraged example to show the relation how to do better.

Reviewer: I think, we should remove the section. If we leave it, also getSiteLanguage(), getContext(), etc. should be documented. But there were introduced with v9 AFAIK, so there were better APIs in place already.

Author: I removed this section. It is true. It is not directly related. We already have a page id and a language id example.

.. include:: /Includes.rst.txt
.. index:: TSFE; TypoScriptFrontendController
.. _tsfe:

====
TSFE
====

.. todo::

    There are still several public properties / methods in TSFE. Unclear if
    usage should be generally discouraged or only where marked as @internal /
    protected.

What is TSFE?
=============

> TSFE is short for :php:`TypoScriptFrontendController`, a class which exists in the

Author: done

system extension EXT:frontend.

As the name implies: A responsibility of TypoScriptFrontendController
is page rendering. It also handles reading from and writing to the page cache.
For more details it is best to look in the source code.

There are at least 3 contexts in which the term TSFE is used:

> 1.  It was and is available as global array :php:`$GLOBALS['TSFE']` in PHP.

Author: done

2.  TypoScript function :ref:`TSFE <t3tsref:data-type-gettext-tsfe>` which can
    be used to access public properties in TSFE.

> 3.  (deprecated since v9.5) Page TSconfig :typoscript:`TSFE.constants`.

Author: done

> Focusing on the PHP part (as the TypoScript part is covered in the TypoScript

Author: done.

    rid of this in the future.

:php:`$GLOBALS['TSFE']` is not available in all contexts. In particular, it is
only available in frontend contexts, not in the backend or cli.

Initializing :php:`$GLOBALS['TSFE']` in the Backend is sometimes done in code
examples found online. TSFE is not initialized in the backend context by the
core (and there is usually no need to do this).

From the PHP documentation:

    As of PHP 8.1.0, $GLOBALS is now a read-only copy of the global symbol table.
    That is, global variables cannot be modified via its copy.

https://www.php.net/manual/en/reserved.variables.globals.php

Howtos
======

.. add howtos for common use cases, remove the not recommended, deprecated
.. way if there is a better way

Following are some examples which use TSFE and alternatives to using TSFE,
where available:

-------------

Access the :php:`ContentObjectRenderer` (often referred to as "cObj"):

.. code-block:: php

    $GLOBALS['TSFE']->cObj

Can be done as follows in Extbase controllers:

.. code-block:: php

    $contentObj = $this->configurationManager->getContentObject();

-------------

Access the current page id:

.. code-block:: php

>     $GLOBALS['TSFE']->id

Author: done

Can be done using the ContentObjectRenderer (see previous example):

.. code-block:: php

    $pageId = $contentObj->data['pid'];

-------------

Access language settings is now done using the
:ref:`language aspect <context_api_aspects_language>`.

> Get the request language of the current page as integer:

Reviewer: Additionally, use the language request attribute: https://docs.typo3.org/m/typo3/reference-coreapi/11.5/en-us/ApiOverview/RequestLifeCycle/RequestAttributes/Language.html

Author: done.

.. code-block:: php

    $context->getPropertyFromAspect('language', 'id');

--------------

Accessing information about Frontend and Backend users.

.. code-block:: php

>     $feUser = $GLOBALS['TSFE']->fe_user;

Reviewer: Request attribute is another possibility (as this is mostly available in Extbase controllers): https://docs.typo3.org/m/typo3/reference-coreapi/11.5/en-us/ApiOverview/RequestLifeCycle/RequestAttributes/FrontendUser.html

Author: done. Thanks again for the link.

Some information via frontend and backend users con be obtained via the
:ref: `user aspect <context_api_aspects_user>`. For example:

.. code-block:: php

    // return whether a frontend user is logged in
    $context->getPropertyFromAspect('frontend.user', 'isLoggedIn');

------------------

To get the base URL of the current site, use site configuration.

For example, to get the site via the current page id:

.. code-block:: php

>     $siteFinder = GeneralUtility::makeInstance(SiteFinder::class);

Reviewer: One should use either dependency injection to get the SiteFinder or just use the site request attribute: https://docs.typo3.org/m/typo3/reference-coreapi/11.5/en-us/ApiOverview/RequestLifeCycle/RequestAttributes/Site.html

Background: We would like to get rid of GeneralUtility::makeInstance() and use DI as best practise wherever possible.

Author: done. Thanks for the link.

    getSiteByPageId($pageId);

Hooks and events
================

.. todo::

    Add overview of hooks and events that are dispatched / called in TSFE

Properties and functions
========================

.. todo::
    here some still used properties and functions can be listed and explained,
    e.g. fe_user, id etc.

> Deprecated functionality (PHP)

Reviewer: Do we really need this part? And if, please only add those which are available in v12 and maybe v11.

Author: The reason is the information is currently spread over several changelogs. You could also look in the source code, but then you don't have the context and the alternatives. 

The section is meant to be removed entirely in the future when there is less activity in that area. The reason I also added the older ones - we did not have this page in the past. I find it helpful to have one overview. (I realize we usually do not add notes if somethings was deprecated several versions ago. This is an exception because the page now only exists in main, and maybe 11.5).

The other reason - one motivation was to "catch" search engine results. If you search for one of the deprecated options, you are directed to pages not on docs.typo3.org or to the changelogs. I would like to direct people to this main page where the information is not fragmented and where they can find information for the specific TYPO3 version in the future (and then use the version selector if they were directed to an older version).

I find this especially important - to direct people to main docs pages - where there is lots of activity and changes and possibly outdated snippets exist in multiple places.

Reviewer: What about this: For this PR we leave them as-is, merge it back to 11.5 and 10.4 and then clean up main (v12) for the outdated ones not related to v12 in a follow-up PR. This way we don't clutter up the recent documentation.

Additionally, we should not merge to 11.5 and 10.4 information for newer versions.

Author: Chris, that sounds good. Removing the v12 changelogs from the 11 and 10 backports should not be too much work, also removing the older changelogs from the main).

Author: I removed the deprecated section and the empty sections for now. It is a bit much for one change and reviewing for other branches as well. That could be added as followup patch.
